src/compliment/sources/class_members.clj

```diff
@@ -209,43 +274,67 @@
 ;; ## Static members
 
 (defn static-member-symbol?
-  "Tests if prefix looks like a static member symbol, returns parsed parts."
-  [x]
+  "Tests if `x` looks like a static member symbol, returns parsed parts."
+  [^String x]
   (re-matches #"([^\/\:\.][^\:]*)\/(.*)" x))
 
-(def static-members-cache
-  "Stores cache of all static members for every class."
+(def ^:private class-members-cache
+  "Members by class"
   (atom {}))
 
-(defn populate-static-members-cache
-  "Populates static members cache for a given class."
+(defn- populate-class-members-cache
+  "Populates qualified methods cache for a given class."
   [^Class class]
-  (swap! static-members-cache assoc class
-         (reduce (fn [cache ^Member c]
-                   (if (static? c)
-                     (update cache (.getName c) (fnil conj []) c)
-                     cache))
-                 {} (concat (.getMethods class) (.getFields class)))))
+  (let [member->cache-key #(get {:constructor "new"} %2 (.getName ^Member %1))
+        methods-by-type   (group-by (comp {true :static-method false :method} static?)
+                                    (.getMethods class))
+        members-by-type   (merge methods-by-type
+                                 {:static-field (.getFields class)
+                                  :constructor  (.getConstructors class)})]
+    (swap! class-members-cache assoc class
+           (reduce (fn [cache [type members]]
+                     (reduce
+                      (fn [acc ^Member m]
+                        (update acc (member->cache-key m type) (fnil conj (with-meta [] {:type type})) m))
+                      cache members))
+                   {}
+                   members-by-type))))
```

Comment:
This whole thing be more efficient if you do two `reduce`s, once over the members and then, optionally, over the constructors.